xc7/utils/prjxray_form_channels.py

```diff
@@ -1429,37 +1432,126 @@ def traverse_pip(conn, wire_in_tile_pkey):
     return None
 
 
+def update_wire_in_tile_types(
+        cur, write_cur, empty_tile_type_pkey, wire_in_tile_pkey_to_update
+):
+    # Check if each wire in tile pkey can point to one logical tile type.
+    wire_in_tile_pkey_to_tile_type_pkey = {}
+
+    for tile_type_pkey, wire_in_tile_pkeys in wire_in_tile_pkey_to_update.items(
+    ):
+        for wire_in_tile_pkey in wire_in_tile_pkeys:
+            if wire_in_tile_pkey not in wire_in_tile_pkey_to_tile_type_pkey:
+                wire_in_tile_pkey_to_tile_type_pkey[wire_in_tile_pkey] = set()
+
+            wire_in_tile_pkey_to_tile_type_pkey[wire_in_tile_pkey].add(
+                tile_type_pkey
+            )
+
+    wire_in_tile_pkey_to_updates = list(
+        wire_in_tile_pkey_to_tile_type_pkey.keys()
+    )
+    wire_in_tile_pkey_updates = []
+    for wire_in_tile_pkey in wire_in_tile_pkey_to_updates:
+        cur.execute(
+            "SELECT phy_tile_type_pkey FROM wire_in_tile WHERE pkey = ?",
+            (wire_in_tile_pkey, )
+        )
+        phy_tile_type_pkey = cur.fetchone()[0]
+        if phy_tile_type_pkey == empty_tile_type_pkey:
+            del wire_in_tile_pkey_to_tile_type_pkey[wire_in_tile_pkey]
+            continue
+
+        tile_type_pkeys = wire_in_tile_pkey_to_tile_type_pkey[wire_in_tile_pkey
+                                                              ]
+        assert len(tile_type_pkeys) == 1, (wire_in_tile_pkey, tile_type_pkeys)
+
+        wire_in_tile_pkey_updates.append(
+            (list(tile_type_pkeys)[0], wire_in_tile_pkey)
+        )
+
+    write_cur.executemany(
+        """
+    UPDATE wire_in_tile
+    SET tile_type_pkey = ?
+    WHERE pkey = ?;""", wire_in_tile_pkey_updates
+    )
+
+
 def create_vpr_grid(conn):
     """ Create VPR grid from prjxray grid. """
     cur = conn.cursor()
     cur2 = conn.cursor()
 
+    slice_types = {}
+
     write_cur = conn.cursor()
     write_cur.execute("""BEGIN EXCLUSIVE TRANSACTION;""")
 
-    # Insert synthetic tile types for CLB sites.
-    write_cur.execute('INSERT INTO tile_type(name) VALUES ("SLICEL");')
-    slicel_tile_type_pkey = write_cur.lastrowid
-
-    write_cur.execute('INSERT INTO tile_type(name) VALUES ("SLICEM");')
-    slicem_tile_type_pkey = write_cur.lastrowid
+    # Insert synthetic tile types for CLB sites and iopad sites.
+    for name in [
+            'SLICEL',
+            'SLICEM',
+            'LIOPAD_M',
```

Comment:
Do we need to have different IOPAD tiles for L and R ?

Reply:
We don't need logical placement types for L and R, but the routing interface for L and R are different.